### PR TITLE
Prometheus: Implement stepMode for alerting queries

### PR DESCRIPTION
**Pull request by ivanahuckova**
**What this PR does / why we need it**:
This PR is continuation to [Prometheus: Add functionality to specify desired step interval in dashboards panels PR](https://github.com/grafana/grafana/pull/36422) that adds option to select **MIN/MAX/EXACT step for Prometheus on frontend**. Before, we had only MIN step option. This PR implements the functionality on backend. 

At first, I have tried to keep the change just within the `prometheus.go` but it was very messy and hard to read. Therefore after discussion in this PR, we have decided to implement the change in `interval.go`.

### CODE REVIEW

**Review by olbo98 — APPROVED** [review]
As far as i can tell the logic seems to work. 🎉  i'm not that familiar with golang though so i've probably missed something. @aocenas should probably review this as well before merging

**Comment by gabor** [comment]
@ivanahuckova i tested it and it works fine, but there is a case where the result might be wrong, could you doublecheck this case please?
- add a prom-query to a dashboard
- time-range last-5-minutes
- step set to "Max", no duration set for step
- do a query, the STEP sent to prometheus is 15sec (i think it is because of the scrape-interval (min-interval) defined in the prom-data-source-by-default)
- now do an old-alert, step STEP sent to prometheus is 0.2

File: pkg/tsdb/prometheus/prometheus.go [hunk 1]
```diff
@@ -141,14 +141,36 @@ func (e *PrometheusExecutor) parseQuery(dsInfo *models.DataSource, query plugins
...
+				// Otherwise use max intervalValue
+			} else {
+				adjustedIntervalValue = intervalValue
+			}
+		} else if hasQueryInterval && stepMode == "exact" {
+			// With exact stepMode - use passed interval value
+			adjustedIntervalValue = intervalValue
+		} else {
+			// With default stepMode - use interval calculator to get ideal interval based on time range and passed value
+			adjustedIntervalValue = e.intervalCalculator.Calculate(*query.TimeRange, intervalValue).Value
```
Comment (aocenas) [inline pkg/tsdb/prometheus/prometheus.go:170]: This was a bit confusing to me. Basically we have branches that does max/exact and then one that does both min and "auto". I thought this was missing one `else if stepMode == "min"` branch. Would it make sense to just keep this inside `Calculate` function? So it would be `Calculate(timerange, interval, stepMode)` or something like that and all this logic would be there?

Comment (gabor) [reply pkg/tsdb/prometheus/prometheus.go:170]: @aocenas  the `Calculate` function is a shared function used by many datasources (which do not have the step-mode), so maybe create a "prometheusCalculate" function that wraps it and does it's thing?

Comment (aocenas) [reply pkg/tsdb/prometheus/prometheus.go:170]: well I guess we can have some defaults there or something? Right now it has a default stepMode == min.

Comment (aocenas) [reply pkg/tsdb/prometheus/prometheus.go:170]: It would need different signature but I assume this does not have to be backward compatible does it?

Comment (ivanahuckova) [reply pkg/tsdb/prometheus/prometheus.go:170]: As none of the other data sources support other interval type than `min`, and as far as I know there is no plan to support it, I tried to do minimal changes that would affect only `prometheus.go`. Moreover, we plan to switch prometheus to backend data source soon, which means that most of the code is going to change/improve/grow. I didn't want to do changes that would affect other data sources that might change again. So I looked at this as temporary smallest-impact solution that will be refactored with all of the incoming changes. 

@aocenas  Do you think it makes sense to update Calculate method now, or would you be okay with this solution that is going to be changed and cleaned up when we move prometheus to be backend data source?

Comment (ivanahuckova) [reply pkg/tsdb/prometheus/prometheus.go:170]: This also still need an update cause in the frontend PR we are always using safe interval, if calculated interval is unsafe. Therefore we need to implement this also here. Will do it after we decide on how to progress.

Comment (aocenas) [reply pkg/tsdb/prometheus/prometheus.go:170]: I don't think this needs to affect other data sources. I mean yeah if the signature changes than there will be change but I don't think that is significant. Ie if have function that calculates interval that right now is specialized to compute interval from range and using also minInterval, generalizing it to function that calculates interval and takes different limit values "min/max/exact" still makes sense to me from stand point of how the code is structured and what the current abstraction is. Which data source uses which limits is sort of inconsequential, we have a function to calculate the interval and different data sources will use it in different way.

I mean this is not a strong difference/preference on my side but such code structure makes more sense to me, especially when looking at this particular set of `if else` where instead of what I expected min/max/exact/default cases it is max/exact/(min+default).

Comment (gabor) [reply pkg/tsdb/prometheus/prometheus.go:170]: one complication might be that it seems datasources are migrating to the "new api", and datasources on old-api use https://github.com/grafana/grafana/blob/main/pkg/tsdb/interval/interval.go#L57 and datasources on new-api use https://github.com/grafana/grafana/blob/main/pkg/tsdb/calculator.go#L56

Comment (gabor) [reply pkg/tsdb/prometheus/prometheus.go:170]: i do agree that the code with the IFs is hard to follow though :-(

Comment (ivanahuckova) [reply pkg/tsdb/prometheus/prometheus.go]: I have refactored it here: https://github.com/grafana/grafana/pull/36796/commits/045db09063fd26153b530b9b3f8ce825097d5c12. It definitely seems more readable. Also, prometheus currently use old api, so I refactored Calculate method in old api in `interval.go`. I merge main and fix merge conflicts for the ddata sources that use new api from `calculator.go`. Can you check it out.

Comment (ivanahuckova) [reply pkg/tsdb/prometheus/prometheus.go]: I have also added safe interval check as it is on frontend.

File: pkg/tsdb/interval/interval.go [hunk 2]
```diff
@@ -53,16 +54,35 @@ func (i *Interval) Milliseconds() int64 {
...
+	case "min":
+		if calculatedInterval < interval {
+			return Interval{Text: FormatDuration(interval), Value: interval}
+		}
+	case "max":
+		if calculatedInterval > interval {
+			return Interval{Text: FormatDuration(interval), Value: interval}
+		}
+	case "exact":
+		return Interval{Text: FormatDuration(interval), Value: interval}
```
Comment (gabor) [inline pkg/tsdb/interval/interval.go:73]: i recommend putting in a `default` case that explodes 💥 , to catch cases when people make a typo in that string, maybe even `panic()` there, like here for example:
https://github.com/grafana/grafana/blob/main/pkg/tsdb/postgres/tlsmanager.go#L92-L101

Comment (gabor) [reply pkg/tsdb/interval/interval.go:73]: hmm.. or if not `panic()` then at least logging something to the error-log (i'm not sure what is the policy on `panic()`, maybe check with backend-team?)

File: docs/sources/datasources/prometheus.md [hunk 3]
```diff
@@ -42,19 +42,19 @@ Open a graph in edit mode by clicking the title > Edit (or by pressing `e` key w
...
+| `Query expression` | Prometheus query expression, check out the [Prometheus documentation](http://prometheus.io/docs/querying/basics/).                                                                                                                                                                                                                          |
+| `Legend format`    | Controls the name of the time series, using name or pattern. For example `{{hostname}}` is replaced with the label value for the label `hostname`.                                                                                                                                                                                          |
+| `Step`             | Use 'Minimum' or 'Maximum' step mode to set the lower or upper bounds respectively on the interval between data points. For example, set "minimum 1h" to hint that measurements were not taken more frequently. Use the 'Exact' step mode to set an exact interval between data points. `$__interval` and `$__rate_interval` are supported. |
```
Comment (achatterjee-grafana) [inline docs/sources/datasources/prometheus.md:49]:
```suggestion
| `Step`             | Use 'Minimum' or 'Maximum' step mode to set the lower or upper bounds respectively on the interval between data points. For example, set "minimum 1h" to hint that measurements are not frequent (taken hourly). Use the 'Exact' step mode to set a precise interval between data points. `$__interval` and `$__rate_interval` are supported. |
```

File: docs/sources/datasources/prometheus.md [hunk 4]
```diff
@@ -42,19 +42,19 @@ Open a graph in edit mode by clicking the title > Edit (or by pressing `e` key w
...
 
-| `Resolution`        | `1/1` sets both the `$__interval` variable and the [`step` parameter of Prometheus range queries](https://prometheus.io/docs/prometheus/latest/querying/api/#range-queries) such that each pixel corresponds to one data point. For better performance, lower resolutions can be picked. `1/2` only retrieves a data point for every other pixel, and `1/10` retrieves one data point per 10 pixels. Note that both _Min time interval_ and _Step_ limit the final value of `$__interval` and `step`. |
+| `Resolution` | `1/1` sets both the `$__interval` variable and the [`step` parameter of Prometheus range queries](https://prometheus.io/docs/prometheus/latest/querying/api/#range-queries) such that each pixel corresponds to one data point. For better performance, lower resolutions can be picked. `1/2` only retrieves a data point for every other pixel, and `1/10` retrieves one data point per 10 pixels. Note that both _Min time interval_ and _Step_ limit the final value of `$__interval` and `step`. |
```
Comment (achatterjee-grafana) [inline docs/sources/datasources/prometheus.md:51]:
```suggestion
| 	`Resolution` | `1/1` sets both the `$__interval` variable and the [`step` parameter of Prometheus range queries](https://prometheus.io/docs/prometheus/latest/querying/api/#range-queries) such that each pixel corresponds to one data point. For better performance, you can pick lower resolutions. `1/2` only retrieves a data point for every other pixel, and `1/10` retrieves one data point per 10 pixels. Both _Min time interval_ and _Step_ limit the final value of `$__interval` and `step`. |
```

File: docs/sources/datasources/prometheus.md [hunk 5]
```diff
@@ -42,19 +42,19 @@ Open a graph in edit mode by clicking the title > Edit (or by pressing `e` key w
...
-| `Exemplars`         | Run and show exemplars in the graph.                                                                                                                                                                                                                                                                                                                                                                                                                                                                      |
+| `Metric lookup` | Search for metric names in this input field. |
+| `Format as` | Switch between `Table`, `Time series`, or `Heatmap`. `Table` will only work in the Table panel. `Heatmap` is suitable for displaying metrics of the Histogram type on a Heatmap panel. Under the hood, it converts cumulative histograms to regular ones and sorts series by the bucket bound. |
```
Comment (achatterjee-grafana) [inline docs/sources/datasources/prometheus.md:54]:
```suggestion
| `Format as` | 	You can switch between `Table` `Time series` or `Heatmap` options. The `Table` option works only in the Table panel. `Heatmap` displays metrics of the Histogram type on a Heatmap panel. Under the hood, it converts cumulative histograms to regular ones and sorts series by the bucket bound. |
```

File: docs/sources/datasources/prometheus.md [hunk 6]
```diff
@@ -42,19 +42,19 @@ Open a graph in edit mode by clicking the title > Edit (or by pressing `e` key w
...
+| `Instant` | Perform an "instant" query, to return only the latest value that Prometheus has scraped for the requested time series. Instant queries return results much faster than normal range queries. Use them to look up label sets. |
```
Comment (achatterjee-grafana) [inline docs/sources/datasources/prometheus.md:55]:
```suggestion
| `Instant` | Perform an "instant" query to return only the latest value that Prometheus has scraped for the requested time series. Instant queries can return results much faster than normal range queries. Use them to look up label sets. |
```

File: docs/sources/datasources/prometheus.md [hunk 7]
```diff
@@ -42,19 +42,19 @@ Open a graph in edit mode by clicking the title > Edit (or by pressing `e` key w
...
+| `Min time interval` | This value multiplied by the denominator from the _Resolution_ setting sets a lower limit to both the `$__interval` variable and the [`step` parameter of Prometheus range queries](https://prometheus.io/docs/prometheus/latest/querying/api/#range-queries). Defaults to _Scrape interval_ as set in the data source options. |
```
Comment (achatterjee-grafana) [inline docs/sources/datasources/prometheus.md:56]:
```suggestion
| `Min time interval` | This value multiplied by the denominator from the _Resolution_ setting sets a lower limit to both the `$__interval` variable and the [`step` parameter of Prometheus range queries](https://prometheus.io/docs/prometheus/latest/querying/api/#range-queries). Defaults to _Scrape interval_ as specified in the data source options. |
```

**Review by achatterjee-grafana — COMMENTED** [review]
Added copy-edit suggestions.

**Comment by ivanahuckova** [comment]
@aocenas The Prometheus frontend step PR was merged (milestone 8.2) and I have implemented feedback. Could you please check it again? 🙂

**Comment by ivanahuckova** [comment]
@marefr @idafurjes If you would have time, could you please review this one. Mainly the change in `interval.go` as it affects also other data sources.

**Review by gabor — APPROVED** [review]
LGTM! works fine, code is ok 👍

**Review by achatterjee-grafana — APPROVED** [review]

**Comment by achatterjee-grafana** [comment]
Need backport labels. Either for 8.1.x or for both 8.1.x and 8.0.x. Thanks.

**Comment by ivanahuckova** [comment]
>Need backport labels. Either for 8.1.x or for both 8.1.x and 8.0.x. Thanks.

I've added 8.2 milestone

**Review by aocenas — APPROVED** [review]

**Comment by idafurjes** [comment]
Currently we are implementing SDK contracts, for that implementation we use the Calculate method from calculator.go, which is already used by some datasources, as you can see here #37358 Dimitris is already using the new Calculate for Prometheus, if this change is needed only for Prometheus maybe it would make sense to implement it only in calculator.go. WDYT?

File: pkg/tsdb/calculator.go [hunk 8]
```diff
@@ -53,16 +61,28 @@ func (i *Interval) Milliseconds() int64 {
...
+		}
+	case Max:
+		if calculatedIntrvl > intrvl {
+			return Interval{Text: interval.FormatDuration(intrvl), Value: intrvl}
+		}
+	case Exact:
+		return Interval{Text: interval.FormatDuration(intrvl), Value: intrvl}
+
+	default:
+		panic(fmt.Sprintf("Unrecognized intervalMode %v", intervalMode))
```
Comment (idafurjes) [inline pkg/tsdb/calculator.go:82]: maybe instead of panicking here we could return (Interval, error) and here we could return empty Interval struct and an error that the intervalMode is invalid

File: pkg/tsdb/interval/interval.go (hunk 2 above)
Comment (idafurjes) [reply pkg/tsdb/interval/interval.go:73]: I wouldn't panic, I would log the error or return it

File: pkg/tsdb/interval/interval.go [hunk 9]
```diff
@@ -54,16 +55,37 @@ func (i *Interval) Milliseconds() int64 {
...
+
+	default:
+		panic(fmt.Sprintf("Unrecognized intervalMode %v", intervalMode))
```
Comment (idafurjes) [inline pkg/tsdb/interval/interval.go:76]: As I mentioned I would return the error or log it

File: pkg/tsdb/calculator.go (hunk 8 above)
Comment (ivanahuckova) [reply pkg/tsdb/calculator.go]: 👍 makes sense. Updated in https://github.com/grafana/grafana/pull/36796/commits/c42ebe7d3f2b71d385e26fcac48de5699df70a03

File: pkg/tsdb/interval/interval.go (hunk 9 above)
Comment (ivanahuckova) [reply pkg/tsdb/interval/interval.go]: Updated in https://github.com/grafana/grafana/pull/36796/commits/c42ebe7d3f2b71d385e26fcac48de5699df70a03

File: pkg/tsdb/interval/interval_test.go [hunk 10]
```diff
@@ -15,19 +15,51 @@ func TestIntervalCalculator_Calculate(t *testing.T) {
...
+		{"from 30m to now", plugins.NewDataTimeRange("30m", "now"), "exact", "1ms"},
+		{"from 24h to now", plugins.NewDataTimeRange("24h", "now"), "min", "1m"},
+		{"from 24h to now", plugins.NewDataTimeRange("24h", "now"), "max", "1ms"},
+		{"from 24h to now", plugins.NewDataTimeRange("24h", "now"), "exact", "1ms"},
 	}
 
 	for _, tc := range testCases {
 		t.Run(tc.name, func(t *testing.T) {
-			interval := calculator.Calculate(tc.timeRange, time.Millisecond*1)
+			interval := calculator.Calculate(tc.timeRange, time.Millisecond*1, tc.intervalMode)
```
Comment (idafurjes) [inline pkg/tsdb/interval/interval_test.go:39]:
```suggestion
			interval, err := calculator.Calculate(tc.timeRange, time.Millisecond*1, tc.intervalMode)
			require.Nil(t, err)
```

File: pkg/tsdb/calculator_test.go [hunk 11]
```diff
@@ -15,19 +15,53 @@ func TestIntervalCalculator_Calculate(t *testing.T) {
...
+		{"from 30m to now", backend.TimeRange{From: timeNow, To: timeNow.Add(30 * time.Minute)}, Exact, "1ms"},
+		{"from 1h to now", backend.TimeRange{From: timeNow, To: timeNow.Add(1440 * time.Minute)}, Min, "1m"},
+		{"from 1h to now", backend.TimeRange{From: timeNow, To: timeNow.Add(1440 * time.Minute)}, Max, "1ms"},
+		{"from 1h to now", backend.TimeRange{From: timeNow, To: timeNow.Add(1440 * time.Minute)}, Exact, "1ms"},
 	}
 
 	for _, tc := range testCases {
 		t.Run(tc.name, func(t *testing.T) {
-			interval := calculator.Calculate(tc.timeRange, time.Millisecond*1)
+			interval := calculator.Calculate(tc.timeRange, time.Millisecond*1, tc.intervalMode)
```
Comment (idafurjes) [inline pkg/tsdb/calculator_test.go:39]:
```suggestion
			interval, err := calculator.Calculate(tc.timeRange, time.Millisecond*1, tc.intervalMode)
			require.Nil(t, err)
```

**Review by idafurjes — APPROVED** [review]
both calculator.go and interval.go look good to me. great job :)
